src/app/store/mod.rs

```diff
@@ -493,11 +416,111 @@ impl<S: ProvableStore> ProvableStore for RevertibleStore<S> {
     }
 }
 
+/// A trait that defines how types are decoded/encoded.
+pub(crate) trait Codec {
+    type Type;
+    type Encoded: AsRef<[u8]>;
+
+    fn encode(d: &Self::Type) -> Option<Self::Encoded>;
+
+    fn decode(bytes: &[u8]) -> Option<Self::Type>;
+}
+
+/// A JSON codec that uses `serde_json` to encode/decode as a JSON string
+#[derive(Clone)]
+pub(crate) struct JsonCodec<T>(PhantomData<T>);
+
+impl<T> Codec for JsonCodec<T>
+where
+    T: Serialize + DeserializeOwned,
+{
+    type Type = T;
+    type Encoded = String;
+
+    fn encode(d: &Self::Type) -> Option<Self::Encoded> {
+        serde_json::to_string(d).ok()
+    }
+
+    fn decode(bytes: &[u8]) -> Option<Self::Type> {
+        let json_string = String::from_utf8(bytes.to_vec()).ok()?;
+        serde_json::from_str(&json_string).ok()
+    }
+}
+
+/// A Protobuf codec that uses `prost` to encode/decode
+#[derive(Clone)]
+pub(crate) struct ProtobufCodec<T, R> {
+    domain_type: PhantomData<T>,
+    raw_type: PhantomData<R>,
+}
+
+impl<T, R> Codec for ProtobufCodec<T, R>
+where
+    T: Into<R> + Clone,
+    R: TryInto<T> + Default + prost::Message,
+{
+    type Type = T;
+    type Encoded = Vec<u8>;
+
+    fn encode(d: &Self::Type) -> Option<Self::Encoded> {
+        let r = d.clone().into();
+        Some(r.encode_to_vec())
+    }
+
+    fn decode(bytes: &[u8]) -> Option<Self::Type> {
+        let r = R::decode(bytes).ok()?;
+        r.try_into().ok()
+    }
+}
+
+/// The `TypedStore` provides methods to treat the data stored at given store paths as given Rust types.
+///
+/// It is designed to be aliased for each concrete codec. For example,
+/// ```rust
+/// type CandyStore<S, K, V> = TypedStore<S, K, CandyCodec<V>>;
+/// ```
+#[derive(Clone)]
+pub(crate) struct TypedStore<S, K, C> {
```

Comment:
I see a few problems with this strategy to assign types to a store:
1. There is no type-safe guarantee when accessing a store path that the data there has the type you're expecting.
    + Standard paths (e.g. in IBC ICS-24) at least should be type safe
2. The abstraction we're looking for really is a "typed path" as opposed to a "typed store"; that is, a store has many paths, each of which can hold values of different types.
    + I think a symptom of this is the fact that methods in the `ibc` module do the decoding manually as opposed to using the `TypedStore` abstraction (e.g. `channel_end()`). Specifically in that case, `self.store` is a `SharedStore`, and there is no one `TypedStore`, as currently implemented, that would be appropriate for all methods.

Reply:
https://github.com/informalsystems/basecoin-rs/pull/44/commits/35285743a9d9c42dc99144f1a65d535b037a3ffa